### DISCUSSION TITLE
Adding IPv6 Support

### DISCUSSION THREAD

**Discussion by anjohnson** [epics-base/epics-base#186] [Misc]
This area is intended for technical discussions about converting EPICS software to support IPv6.

**Comment by anjohnson** [comment]
### Portable Interfaces to Network Sockets

From an email by @mdavidsaver, discussing the APIs currently provided in osiSock.h by libCom:

> Early on in developing PVXS I looked at what would be necessary to support `AF_INET6`. The main obstacle I found was the widely used `union osiSockAddr`.  A lot of code uses `sizeof(union osiSockAddr)` as the length argument along with `struct sockaddr*`
>
> Simply adding `struct sockaddr_in6` to the union will resulting in existing code passing a `struct sockaddr_in*` and `sizeof(struct sockaddr_in6)`.  This works fine on Linux and Windows, but fails on BSD derivatives (including OSX). There it seems that the `socklen` argument must exactly match the `AF_*` family, and probably the magic `sa_len` member as well.
>
> As breaking all existing user code on OSX and RTEMS would make us few friends, this simple path forward seems to be closed to us.

**Comment by anjohnson** [comment]
### IP Addresses in Protocols

@tboegi wrote:

> A socket opened with AF_INET6 can handle IPV4 traffic. The underlying OS will "map" the IPv4 addresses into IPv6, look for `IN6_IS_ADDR_V4MAPPED`.
> `192.168.1.1` will be `[0:0:0:0:0:FFFF:C0A0:0101]` (or similar)
> So if we really, really don't want any IPv4 traffic at all, not even localhost, we probably need to filter out the traffic on application layer.

@mdavidsaver replied:

> If we go the route of having a different, longer, message format with IPV6 this would help when receiving, but maybe not when sending.
>
> fyi. With CA, the address encoded in search replies is almost always the special `255.255.255.255` to signify that the recipient should use use the apparent source address (cf. `recvfrom()` ).
> 
> One option would be to make this mandatory for IPv6 traffic.  This would prevent the usage of "name servers".  Though imo. this would be replaced by either a gateway, or use of multicast.

**Comment by anjohnson** [comment]
### Switching from Broadcast to Multicast

**Reply by anjohnson** [reply]
For IPv4 addresses, [RFC5771](https://datatracker.ietf.org/doc/html/rfc5771) says:
```
   In general, due to the relatively small size of the IPv4 multicast
   address space, further assignment of IPv4 multicast address space is
   recommended only in limited circumstances.  Specifically, the IANA
   should only assign addresses in those cases where:

      - the dynamic selection Session Description Protocol/Session
        Announcement Protocol (SDP/SAP);

      - GLOP (not an acronym);

      - Source-Specific Multicast (SSM); or

      - Administratively Scoped address spaces cannot be used.
```
so we're probably out of luck getting our own multicast IP address there, but maybe one of the above methods might provide a solution. Sites might need multiple naming domains so a single address might not work there, although I guess a different port number might be usable in that case.

[RFC3307](https://datatracker.ietf.org/doc/html/rfc3307) appears to cover allocating IPv6 multicast addresses. The current list of allocations from IANA is [here](https://www.iana.org/assignments/ipv6-multicast-addresses/ipv6-multicast-addresses.xhtml).

A lot more research is needed first though.

**Reply by anjohnson** [reply]
MAD pointed out that just using a different port won't limit which hosts receive the packets, whereas a different multicast IP ensures that only interested hosts will get them. Apparently Fermilab are successfully using CA with Multicast; PVA support for multicast will be worked on soon.

**Reply by mdavidsaver** [reply]
wrt. RFC5771 options:

- SAP (transport) / SDP (payload)

This would actually be a reasonable option for beacons.  No so for searches.  SAP wishes to [mandate](https://datatracker.ietf.org/doc/html/rfc2974#section-3.1) a 4KB/s limit of per node announcement traffic.  SDP [payloads](https://datatracker.ietf.org/doc/html/rfc4566#section-5) are multi-line text, so processing/parsing is relatively expensive compared to the current CA/PVA search messages.

- GLOP

[RFC3180](https://datatracker.ietf.org/doc/html/rfc3180) lays out a scheme for picking ipv4 mcast addresses in the 233.0.0.0/8 range based on [Autonomous System number](https://www.iana.org/assignments/as-numbers/as-numbers.xhtml).  While these identifiers seem to be mostly of interest to large scale internet routing, there is a range 64512-65534 reserved for private use.

As I read it, we could just start using an address in the private AS range as a default.

Note, RFC3180 only covers IPv4.  So some other solution would be needed for IPv6.

- Source-Specific Multicast

Not an option since a CA/PVA peer don't know the full list of peer IPs.

- Administratively Scoped address spaces

We could also probably just pick an address in this range as a default.

- MASC

RFC3180 includes a reference to [RFC2909](https://datatracker.ietf.org/doc/html/rfc2909) which defines the Multicast Address-Set Claim (MASC) Protocol.  While an intriguing name, the document is listed as Experimental, and a quick search suggests the idea didn't go anywhere.  So I'm not investigating further.

**Comment by mdavidsaver** [comment]
### OS dependent multicast/socket behaviors

Some things I've discovered...

#### On Linux multicast group membership is system wide, not per-socket, by default.

On Linux, I discovered the default, non-compliant, behavior of `IP_MULTICAST_ALL=1` the hard way.  This should be cleared for all UDP sockets to give consistent/sane behavior.  cf. [`man 7 ip`](https://man7.org/linux/man-pages/man7/ip.7.html).

The [winsock](https://docs.microsoft.com/en-us/windows/win32/winsock/ipproto-ip-socket-options) documentation says "Join the socket to the supplied multicast group on the specified interface.".  The *BSDs do the same.

#### Differences for sockets bound to an mcast address

On Linux, it is possible to bind a socket to a multicast address in order to receive only multicasts (eg. no unicasts or broadcasts).  This socket can also be used to send multicasts.

On OSX (and presumably *BSD, RTEMS) this works like Linux for receiving, but this socket can't be used to send multicasts.

With winsock, it isn't possible to bind to an mcast address.


#### RTEMS classic and libbsd provide `IP_ORIGDSTADDR` and `IP_RECVIF` instead of `IP_PKTINFO`

Not strictly multicast behavior, but I'm trying to use these control messages with a socket bound to `0.0.0.0` in lieu of binding many sockets to individual interfaces.  `IP_PKTINFO` includes the destination IP address (ucast, mcast, or bcast).

I don't see a portable and race-free way to use the interface index provided with `IP_PKTINFO` to find the address of the interface through which the packet was received.

**Comment by tboegi** [comment]
I have some prototype-coding here:

- No IPv6 enabled yet
- code compiles under Linux
- No actions for pvAccess yet
- camonitor spits out "CAC: Undecipherable UDP message from 127.0.0.1:43686"
This is not the final version in any way. Comments of any kind welcome.
https://github.com/tboegi/epics-base/tree/210823-Ipv6-osiSocketBind-osiSockAddr46

**Reply by tboegi** [reply]
Some more progress:
https://github.com/tboegi/epics-base/commit/23d9d99be0c486c4e4a9be4431d83d91ec08cf29
https://github.com/epics-base/epics-base/pull/203
